src/BCFKeyValueData.cc

```diff
@@ -292,30 +338,40 @@ Status BCFKeyValueData::import_gvcf(const DataCache* cache,
     Status s;
     unique_ptr<BCFWriter> writer;
     unique_ptr<bcf1_t, void(*)(bcf1_t*)> vt(bcf_init(), &bcf_destroy);
-    S(BCFWriter::Open(writer));
 
+    range bucket(-1, 0, GENOMIC_INTERVAL_LEN);
     int c;
-    int chrom_id = -1;
     for(c = bcf_read(vcf.get(), hdr.get(), vt.get());
         c == 0;
         c = bcf_read(vcf.get(), hdr.get(), vt.get())) {
-        // Moved to the next chromosome, write this key.
-        if ( vt->rid != chrom_id &&
-             writer->get_num_entries() > 0) {
-            S(write_chrom(body_->db, writer.get(), dataset, chrom_id));
+        // Make sure a record is not longer than [GENOMIC_INTERVAL_LEN].
+        // If this is not true, then we need to compensate in the search
+        // routine.
+        assert(vt->rlen <= GENOMIC_INTERVAL_LEN);
+
+        // should we start a new bucket?
+        if (vt->rid != bucket.rid || vt->pos >= bucket.end) {
+            // write old bucket to DB
+            if (writer != NULL && writer->get_num_entries() > 0)
+                S(write_bucket(body_->db, writer.get(), dataset, bucket));
 
             // start a new in-memory chunk
             writer = NULL;
             S(BCFWriter::Open(writer));
+
+            // Move to a new genomic range. Round down the start address to
+            // a natural multiple of the interval length.
+            bucket.rid = vt->rid;
+            bucket.beg = (vt->pos / GENOMIC_INTERVAL_LEN) * GENOMIC_INTERVAL_LEN;
+            bucket.end = bucket.beg + GENOMIC_INTERVAL_LEN;
         }
-        chrom_id = vt->rid;
         S(writer->write(vt.get()));
     }
     if (c != -1) return Status::IOError("reading from gVCF file", filename);
 
-    // close last chromosome
-    if (writer->get_num_entries() > 0) {
-        S(write_chrom(body_->db, writer.get(), dataset, chrom_id));
+    // close last bucket
+    if (writer != NULL && writer->get_num_entries() > 0) {
+        S(write_bucket(body_->db, writer.get(), dataset, bucket));
     }
     writer = NULL;  // clear the memory state
 
```

Comment:
It would be nice to provide atomic commit of all the key-value pairs for a gVCF file. (Again feel free to file as a future todo)